### PR TITLE
[Optimization] Code editor key binding event optimization.

### PR DESCRIPTION
**Pull request by cloverhearts**
### What is this PR for?
zeppelin custom key binding event optimization for Ace Code editor

### What type of PR is it?
Improvement

### Todos
- [x] optimization for key binding and check logic 

### How should this be tested?
For many in the code editor and typing quickly.
The effect is greater when there are many Paragraph.

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 1]
```diff
@@ -665,6 +665,16 @@ angular.module('zeppelinWebApp').controller('ParagraphCtrl', function($scope, $r
       $scope.editor.commands.bindKey('ctrl-.', 'startAutocomplete');
       $scope.editor.commands.bindKey('ctrl-space', null);
 
+      $scope.keyBindingEditorFocusAction = function(moveTarget, scrollValue) {
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: This should be `var`, not $scope

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 2]
```diff
@@ -678,27 +688,23 @@ angular.module('zeppelinWebApp').controller('ParagraphCtrl', function($scope, $r
...
-            if (currentRow === 0) {
-              // move focus to previous paragraph
-              $scope.$emit('moveFocusToPreviousParagraph', $scope.paragraph.id);
-            } else {
-              $scope.scrollToCursor($scope.paragraph.id, -1);
-            }
-          } else if (keyCode === 40 || (keyCode === 78 && e.ctrlKey && !e.altKey)) {  // DOWN
-            numRows = $scope.editor.getSession().getLength();
-            currentRow = $scope.editor.getCursorPosition().row;
-            if (currentRow === numRows - 1) {
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: You forgot this case in your function above

Comment (cloverhearts) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: @corneadoug 
Thanks Giving find.
I will fix.

Comment (cloverhearts) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: @corneadoug 
Modifications completed.
```
if (currentRow === 0 || (currentRow === numRows - 1)) {
```

**Comment by cloverhearts** [comment]
Sorry, i found bug.

**Comment by cloverhearts** [comment]
Bug fixes completed.
I will wait for the reviews.

**Comment by Leemoonsoo** [comment]
Thanks for the patch. 

Could you create two paragraphs and write multiple lines of text (more than 3 lines) in the code editor and try move focus by up/down key?

Cursor skips top/bottom line of text in the code editor.

**Comment by cloverhearts** [comment]
@Leemoonsoo 
Thank you for your opinion.
I'll check

**Comment by cloverhearts** [comment]
@Leemoonsoo 
testrow
This modification has been completed.
If you have any other opinion or find a bug. , please post a comment.
Thank you, have a nice day.

**Comment by Leemoonsoo** [comment]
Tested and LGTM.
Merge to maste and branch-0.6 if there're no more discussions

**Pull request closed without merging**